File: backend/app/api/routes/scoring.py
"""
Scoring API Routes

Endpoints for molecule scoring including:
- ML-readiness
- NP-likeness
- Scaffold extraction
- Drug-likeness (Lipinski, QED, Veber, Ro3, Ghose, Egan, Muegge)
- Safety filters (PAINS, Brenk, NIH, ZINC, ChEMBL)
- ADMET predictions (SAscore, ESOL, Fsp3, CNS MPO, Pfizer/GSK rules)
- Aggregator likelihood prediction
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    pass

from fastapi import APIRouter, Depends, HTTPException, Request

from app.api.routes.validation import extract_molecule_info
from app.core.rate_limit import get_rate_limit_key, limiter
from app.core.security import get_api_key
from app.schemas.scoring import (
    ComparisonRequest,
    MLReadinessBreakdownSchema,
    MLReadinessResultSchema,
    NPLikenessResultSchema,
    RadarAxisSchema,
    RadarComparisonSchema,
    RadarProfileSchema,
    ScaffoldResultSchema,
    ScoringRequest,
    ScoringResponse,
)
from app.services.parser.molecule_parser import MoleculeFormat, parse_molecule
from app.services.scoring import (
    calculate_ml_readiness,
    calculate_np_likeness,
    calculate_radar_comparison,
    extract_scaffold,
)
from app.services.scoring.score_builders import (
    _calculate_admet,
    _calculate_aggregator,
    _calculate_bertz_detail,
    _calculate_bioavailability_radar_route,
    _calculate_boiled_egg_route,
    _calculate_consensus,
    _calculate_druglikeness,
    _calculate_fsp3_detail,
    _calculate_lead_likeness,
    _calculate_ligand_efficiency,
    _calculate_logp_breakdown,
    _calculate_np_breakdown_route,
    _calculate_safety_filters,
    _calculate_salt_inventory,
    _calculate_tpsa_breakdown,
    _ml_dimension_to_schema,
)

router = APIRouter()


@router.post("/score", response_model=ScoringResponse)
@limiter.limit("10/minute", key_func=get_rate_limit_key)
async def score_molecule(
    request: Request,
    body: ScoringRequest,
    api_key: Optional[str] = Depends(get_api_key),
):
    """
    Calculate scores for a molecule.

    Supported scoring types:
    - ml_readiness: ML-readiness score (0-100) with breakdown
    - np_likeness: Natural product likeness score (-5 to +5)
    - scaffold: Murcko scaffold extraction
    - druglikeness: Drug-likeness filters (Lipinski, QED, Veber, Ro3, etc.)
    - safety_filters: Safety alerts (PAINS, Brenk, NIH, ZINC, ChEMBL)
    - admet: ADMET predictions (SAscore, solubility, Pfizer/GSK rules, etc.)
    - aggregator: Aggregator likelihood prediction

    Args:
        body: ScoringRequest with molecule and options

    Returns:
        ScoringResponse with requested scoring results

    Raises:
        HTTPException: If molecule cannot be parsed
    """
    start_time = time.time()

    # IUPAC/common name resolution (try OPSIN then PubChem)
    molecule_input = body.molecule
    if body.format in ("auto", "iupac"):
        from app.services.iupac.converter import detect_input_type, name_to_smiles

        detected = detect_input_type(molecule_input) if body.format == "auto" else "iupac"
        if detected == "iupac" or (detected == "ambiguous" and body.format == "auto"):
            converted, _source = name_to_smiles(molecule_input)
            if converted:
                molecule_input = converted

    # Parse molecule
    format_map = {
        "smiles": MoleculeFormat.SMILES,
        "inchi": MoleculeFormat.INCHI,
        "mol": MoleculeFormat.MOL,
        "iupac": MoleculeFormat.SMILES,
        "auto": None,
    }
    input_format = format_map.get(body.format)

    parse_result = parse_molecule(molecule_input, input_format)

    if not parse_result.success or parse_result.mol is None:
        raise HTTPException(
            status_code=400,
            detail={
                "error": "Failed to parse molecule",
                "errors": parse_result.errors,
                "warnings": parse_result.warnings,
                "format_detected": parse_result.format_detected.value,
            },
        )

    mol = parse_result.mol

    # Extract basic molecule info
    mol_info = extract_molecule_info(mol, body.molecule)

    # Initialize response components
    ml_readiness_result = None
    np_likeness_result = None
    scaffold_result = None
    druglikeness_result = None
    safety_filters_result = None
    admet_result = None
    aggregator_result = None
    consensus_result = None
    lead_likeness_result = None
    salt_inventory_result = None
    ligand_efficiency_result = None
    tpsa_breakdown_result = None
    logp_breakdown_result = None
    bertz_detail_result = None
    fsp3_detail_result = None
    np_breakdown_result = None
    bioavailability_radar_result = None
    boiled_egg_result = None

    # Calculate requested scores
    if "ml_readiness" in body.include:
        ml_result = calculate_ml_readiness(mol)
        ml_readiness_result = MLReadinessResultSchema(
            score=ml_result.score,
            label=ml_result.label,
            color=ml_result.color,
            breakdown=MLReadinessBreakdownSchema(
                structural_quality=_ml_dimension_to_schema(ml_result.breakdown.structural_quality),
                property_profile=_ml_dimension_to_schema(ml_result.breakdown.property_profile),
                complexity_feasibility=_ml_dimension_to_schema(
                    ml_result.breakdown.complexity_feasibility,
                ),
                representation_quality=_ml_dimension_to_schema(
                    ml_result.breakdown.representation_quality,
                ),
            ),
            caveats=ml_result.caveats,
            supplementary=ml_result.supplementary,
            interpretation=ml_result.interpretation,
        )

    if "np_likeness" in body.include:
        np_result = calculate_np_likeness(mol)
        np_likeness_result = NPLikenessResultSchema(
            score=np_result.score,
            interpretation=np_result.interpretation,
            caveats=np_result.caveats,
            details=np_result.details,
        )

    if "scaffold" in body.include:
        scaffold_res = extract_scaffold(mol)
        scaffold_result = ScaffoldResultSchema(
            scaffold_smiles=scaffold_res.scaffold_smiles,
            generic_scaffold_smiles=scaffold_res.generic_scaffold_smiles,
            has_scaffold=scaffold_res.has_scaffold,
            message=scaffold_res.message,
            details=scaffold_res.details,
        )

    if "druglikeness" in body.include:
        druglikeness_result = _calculate_druglikeness(mol)

    if "safety_filters" in body.include:
        safety_filters_result = _calculate_safety_filters(mol)

    if "admet" in body.include:
        admet_result = _calculate_admet(mol)

    if "aggregator" in body.include:
        aggregator_result = _calculate_aggregator(mol)

    if "consensus" in body.include:
        consensus_result = _calculate_consensus(mol)

    if "lead_likeness" in body.include:
        lead_likeness_result = _calculate_lead_likeness(mol)

    if "salt_inventory" in body.include:
        salt_inventory_result = _calculate_salt_inventory(mol)

    if "ligand_efficiency" in body.include:
        ligand_efficiency_result = _calculate_ligand_efficiency(mol)

    if "tpsa_breakdown" in body.include:
        tpsa_breakdown_result = _calculate_tpsa_breakdown(mol)

    if "logp_breakdown" in body.include:
        logp_breakdown_result = _calculate_logp_breakdown(mol)

    if "bertz_detail" in body.include:
        bertz_detail_result = _calculate_bertz_detail(mol)

    if "fsp3_detail" in body.include:
        fsp3_detail_result = _calculate_fsp3_detail(mol)

    if "np_breakdown" in body.include:
        np_breakdown_result = _calculate_np_breakdown_route(mol)

    if "bioavailability_radar" in body.include:
        bioavailability_radar_result = _calculate_bioavailability_radar_route(mol)

    if "boiled_egg" in body.include:
        boiled_egg_result = _calculate_boiled_egg_route(mol)

    execution_time = int((time.time() - start_time) * 1000)

    return ScoringResponse(
        molecule_info=mol_info,
        ml_readiness=ml_readiness_result,
        np_likeness=np_likeness_result,
        scaffold=scaffold_result,
        druglikeness=druglikeness_result,
        safety_filters=safety_filters_result,
        admet=admet_result,
        aggregator=aggregator_result,
        consensus=consensus_result,
        lead_likeness=lead_likeness_result,
        salt_inventory=salt_inventory_result,
        ligand_efficiency=ligand_efficiency_result,
        tpsa_breakdown=tpsa_breakdown_result,
        logp_breakdown=logp_breakdown_result,
        bertz_detail=bertz_detail_result,
        fsp3_detail=fsp3_detail_result,
        np_breakdown=np_breakdown_result,
        bioavailability_radar=bioavailability_radar_result,
        boiled_egg=boiled_egg_result,
        execution_time_ms=execution_time,
    )


@router.post("/score/compare", response_model=RadarComparisonSchema)
@limiter.limit("10/minute", key_func=get_rate_limit_key)
async def compare_molecules(
    request: Request,
    body: ComparisonRequest,
    api_key: Optional[str] = Depends(get_api_key),
):
    """
    Compare multiple molecules using bioavailability radar profiles.

    Accepts up to 10 SMILES strings and returns radar profiles for each,
    plus a drug-like reference profile.

    Args:
        body: ComparisonRequest with list of SMILES

    Returns:
        RadarComparisonSchema with profiles and reference
    """
    result = calculate_radar_comparison(body.smiles_list)

    profiles = [
        RadarProfileSchema(
            smiles=p.smiles,
            axes=[
                RadarAxisSchema(
                    name=a.name,
                    actual_value=a.actual_value,
                    normalized=a.normalized,
                    optimal_min=a.optimal_min,
                    optimal_max=a.optimal_max,
                    in_range=a.in_range,
                    property_name=a.property_name,
                    unit=a.unit,
                )
                for a in p.axes
            ],
            is_reference=p.is_reference,
        )
        for p in result.profiles
    ]

    reference = None
    if result.reference:
        reference = RadarProfileSchema(
            smiles=result.reference.smiles,
            axes=[
                RadarAxisSchema(
                    name=a.name,
                    actual_value=a.actual_value,
                    normalized=a.normalized,
                    optimal_min=a.optimal_min,
                    optimal_max=a.optimal_max,
                    in_range=a.in_range,
                    property_name=a.property_name,
                    unit=a.unit,
                )
                for a in result.reference.axes
            ],
            is_reference=result.reference.is_reference,
        )

    return RadarComparisonSchema(
        profiles=profiles,
        reference=reference,
    )

File: frontend/src/types/scoring.ts
/**
 * Scoring Types
 *
 * TypeScript interfaces for ML-readiness, NP-likeness, scaffold scoring,
 * drug-likeness, safety filters, and ADMET predictions.
 */

/**
 * Breakdown of ML-readiness score components.
 */
export interface MLReadinessBreakdown {
  // Standard descriptors (CalcMolDescriptors - 217 descriptors)
  descriptors_score: number;
  descriptors_max: number;
  descriptors_successful: number;
  descriptors_total: number;

  // Additional descriptors (AUTOCORR2D + MQN)
  additional_descriptors_score: number;
  additional_descriptors_max: number;
  autocorr2d_successful: number;
  autocorr2d_total: number;
  mqn_successful: number;
  mqn_total: number;

  // Fingerprints (7 types)
  fingerprints_score: number;
  fingerprints_max: number;
  fingerprints_successful: string[];
  fingerprints_failed: string[];

  // Size constraints
  size_score: number;
  size_max: number;
  molecular_weight: number | null;
  num_atoms: number | null;
  size_category: 'optimal' | 'acceptable' | 'out_of_range' | 'error' | 'unknown';
}

/**
 * ML-readiness scoring result.
 */
export interface MLReadinessResult {
  score: number;
  breakdown: MLReadinessBreakdown;
  interpretation: string;
  failed_descriptors: string[];
}

/**
 * NP-likeness scoring result.
 */
export interface NPLikenessResult {
  score: number;
  interpretation: string;
  caveats: string[];
  details: Record<string, unknown>;
}

/**
 * Scaffold extraction result.
 */
export interface ScaffoldResult {
  scaffold_smiles: string;
  generic_scaffold_smiles: string;
  has_scaffold: boolean;
  message: string;
  details: Record<string, unknown>;
}

// =============================================================================
// Drug-likeness Types
// =============================================================================

/**
 * Lipinski's Rule of Five results.
 */
export interface LipinskiResult {
  passed: boolean;
  violations: number;
  mw: number;
  logp: number;
  hbd: number;
  hba: number;
  details: Record<string, boolean>;
}

/**
 * QED score results.
 */
export interface QEDResult {
  score: number;
  properties: Record<string, number | string>;
  interpretation: string;
}

/**
 * Veber rules results.
 */
export interface VeberResult {
  passed: boolean;
  rotatable_bonds: number;
  tpsa: number;
}

/**
 * Rule of Three (fragment-likeness) results.
 */
export interface RuleOfThreeResult {
  passed: boolean;
  violations: number;
  mw: number;
  logp: number;
  hbd: number;
  hba: number;
  rotatable_bonds: number;
  tpsa: number;
}

/**
 * Ghose filter results.
 */
export interface GhoseResult {
  passed: boolean;
  violations: number;
  mw: number;
  logp: number;
  atom_count: number;
  molar_refractivity: number;
}

/**
 * Egan filter results.
 */
export interface EganResult {
  passed: boolean;
  logp: number;
  tpsa: number;
}

/**
 * Muegge filter results.
 */
export interface MueggeResult {
  passed: boolean;
  violations: number;
  details: Record<string, boolean>;
}

/**
 * Complete drug-likeness scoring results.
 */
export interface DrugLikenessResult {
  lipinski: LipinskiResult;
  qed: QEDResult;
  veber: VeberResult;
  ro3: RuleOfThreeResult;
  ghose: GhoseResult | null;
  egan: EganResult | null;
  muegge: MueggeResult | null;
  interpretation: string;
}

// =============================================================================
// Safety Filters Types
// =============================================================================

/**
 * Result for a single filter category.
 */
export interface FilterAlertResult {
  passed: boolean;
  alerts: string[];
  alert_count: number;
}

/**
 * ChEMBL structural alerts results.
 */
export interface ChEMBLAlertsResult {
  passed: boolean;
  total_alerts: number;
  bms: FilterAlertResult | null;
  dundee: FilterAlertResult | null;
  glaxo: FilterAlertResult | null;
  inpharmatica: FilterAlertResult | null;
  lint: FilterAlertResult | null;
  mlsmr: FilterAlertResult | null;
  schembl: FilterAlertResult | null;
}

/**
 * Complete safety filter results.
 */
export interface SafetyFilterResult {
  pains: FilterAlertResult;
  brenk: FilterAlertResult;
  nih: FilterAlertResult | null;
  zinc: FilterAlertResult | null;
  chembl: ChEMBLAlertsResult | null;
  all_passed: boolean;
  total_alerts: number;
  interpretation: string;
}

// =============================================================================
// ADMET Types
// =============================================================================

/**
 * Synthetic accessibility score result.
 */
export interface SyntheticAccessibilityResult {
  score: number;
  classification: 'easy' | 'moderate' | 'difficult' | 'unknown';
  interpretation: string;
}

/**
 * ESOL solubility prediction result.
 */
export interface SolubilityResult {
  log_s: number;
  solubility_mg_ml: number;
  classification: 'highly_soluble' | 'soluble' | 'moderate' | 'poor' | 'insoluble' | 'unknown';
  interpretation: string;
}

/**
 * Molecular complexity metrics.
 */
export interface ComplexityResult {
  fsp3: number;
  num_stereocenters: number;
  num_rings: number;
  num_aromatic_rings: number;
  bertz_ct: number;
  classification: 'flat' | 'moderate' | '3d' | 'unknown';
  interpretation: string;
}

/**
 * CNS MPO score result.
 */
export interface CNSMPOResult {
  score: number;
  components: Record<string, number>;
  cns_penetrant: boolean;
  interpretation: string;
}

/**
 * Bioavailability indicators.
 */
export interface BioavailabilityResult {
  tpsa: number;
  rotatable_bonds: number;
  hbd: number;
  hba: number;
  mw: number;
  logp: number;
  oral_absorption_likely: boolean;
  cns_penetration_likely: boolean;
  interpretation: string;
}

/**
 * Pfizer 3/75 Rule result.
 */
export interface PfizerRuleResult {
  passed: boolean;
  logp: number;
  tpsa: number;
  interpretation: string;
}

/**
 * GSK 4/400 Rule result.
 */
export interface GSKRuleResult {
  passed: boolean;
  mw: number;
  logp: number;
  interpretation: string;
}

/**
 * Golden Triangle (Abbott) analysis.
 */
export interface GoldenTriangleResult {
  in_golden_triangle: boolean;
  mw: number;
  logd: number;
  interpretation: string;
}

/**
 * Complete ADMET prediction results.
 */
export interface ADMETResult {
  synthetic_accessibility: SyntheticAccessibilityResult;
  solubility: SolubilityResult;
  complexity: ComplexityResult;
  cns_mpo: CNSMPOResult | null;
  bioavailability: BioavailabilityResult;
  pfizer_rule: PfizerRuleResult | null;
  gsk_rule: GSKRuleResult | null;
  golden_triangle: GoldenTriangleResult | null;
  molar_refractivity: number | null;
  interpretation: string;
}

// =============================================================================
// Aggregator Likelihood Types
// =============================================================================

/**
 * Aggregator likelihood prediction result.
 */
export interface AggregatorLikelihoodResult {
  likelihood: 'low' | 'moderate' | 'high';
  risk_score: number;
  logp: number;
  tpsa: number;
  mw: number;
  aromatic_rings: number;
  risk_factors: string[];
  interpretation: string;
  confidence: number;
  evidence: Array<Record<string, unknown>>;
}

// =============================================================================
// Consensus Drug-Likeness Types
// =============================================================================

/**
 * Per-property result within a rule set.
 */
export interface RuleViolation {
  property: string;
  value: number;
  threshold: string;
  result: 'pass' | 'fail';
}

/**
 * Detail for a single rule set in consensus scoring.
 */
export interface RuleSetDetail {
  name: string;
  passed: boolean;
  violations: RuleViolation[];
}

/**
 * Consensus drug-likeness score across 5 rule sets.
 */
export interface ConsensusScore {
  score: number;
  total: number;
  rule_sets: RuleSetDetail[];
  interpretation: string;
}

// =============================================================================
// Lead-Likeness Types
// =============================================================================

/**
 * Lead-likeness assessment result.
 */
export interface LeadLikeness {
  passed: boolean;
  violations: number;
  properties: Record<string, number>;
  thresholds: Record<string, string>;
  violation_details: RuleViolation[];
}

// =============================================================================
// Salt Inventory Types
// =============================================================================

/**
 * A single fragment from a salt-form molecule.
 */
export interface SaltFragment {
  smiles: string;
  name: string;
  category: string;
  mw: number;
  heavy_atom_count: number;
}

/**
 * Salt/counterion inventory result.
 */
export interface SaltInventory {
  has_salts: boolean;
  parent_smiles: string;
  fragments: SaltFragment[];
  total_fragments: number;
  interpretation: string;
}

// =============================================================================
// Ligand Efficiency Types
// =============================================================================

/**
 * Ligand efficiency result.
 */
export interface LigandEfficiency {
  le: number | null;
  heavy_atom_count: number;
  activity_value: number | null;
  activity_type: string | null;
  proxy_used: boolean;
  interpretation: string;
}

// =============================================================================
// Property Breakdown Types
// =============================================================================

/**
 * Per-atom property contribution.
 */
export interface AtomContribution {
  atom_index: number;
  symbol: string;
  contribution: number;
}

/**
 * Per-functional-group property contribution.
 */
export interface FunctionalGroupContribution {
  group_name: string;
  contribution: number;
  atom_indices: number[];
}

/**
 * TPSA per-atom breakdown.
 */
export interface TPSABreakdown {
  total: number;
  atom_contributions: AtomContribution[];
  functional_group_summary: FunctionalGroupContribution[];
}

/**
 * LogP per-atom breakdown.
 */
export interface LogPBreakdown {
  total: number;
  atom_contributions: AtomContribution[];
  functional_group_summary: FunctionalGroupContribution[];
}

/**
 * Bertz complexity detail.
 */
export interface BertzDetail {
  bertz_ct: number;
  num_bonds: number;
  num_atoms: number;
  num_rings: number;
  num_aromatic_rings: number;
  ring_complexity: number;
  interpretation: string;
}

/**
 * Per-carbon hybridization data.
 */
export interface CarbonHybridization {
  atom_index: number;
  symbol: string;
  hybridization: string;
}

/**
 * Fsp3 per-carbon detail.
 */
export interface Fsp3Detail {
  fsp3: number;
  total_carbons: number;
  sp3_count: number;
  sp2_count: number;
  sp_count: number;
  per_carbon: CarbonHybridization[];
  interpretation: string;
}

// =============================================================================
// NP-Likeness Breakdown Types
// =============================================================================

/**
 * A single NP fragment contribution.
 */
export interface NPFragment {
  smiles: string;
  contribution: number;
  bit_id: number;
  radius: number;
  center_atom_idx: number;
  classification: 'np_characteristic' | 'synthetic_characteristic' | 'neutral';
}

/**
 * NP-likeness fragment breakdown.
 */
export interface NPBreakdown {
  score: number;
  confidence: number;
  fragments: NPFragment[];
  total_fragments: number;
  np_fragment_count: number;
  synthetic_fragment_count: number;
  interpretation: string;
}

// =============================================================================
// Bioavailability Radar & BOILED-Egg Types
// =============================================================================

/**
 * A single axis of the bioavailability radar.
 */
export interface RadarAxis {
  name: string;
  actual_value: number;
  normalized: number;
  optimal_min: number;
  optimal_max: number;
  in_range: boolean;
  property_name: string;
  unit: string;
}

/**
 * Bioavailability radar result.
 */
export interface BioavailabilityRadar {
  axes: RadarAxis[];
  overall_in_range_count: number;
  interpretation: string;
}

/**
 * Ellipse parameters for BOILED-Egg model.
 */
export interface EllipseParams {
  cx: number;
  cy: number;
  a: number;
  b: number;
}

/**
 * BOILED-Egg classification result.
 */
export interface BoiledEgg {
  wlogp: number;
  tpsa: number;
  gi_absorbed: boolean;
  bbb_permeant: boolean;
  region: 'yolk' | 'white' | 'grey';
  gi_ellipse: EllipseParams | null;
  bbb_ellipse: EllipseParams | null;
  interpretation: string;
}

/**
 * A single molecule's radar profile.
 */
export interface RadarProfile {
  smiles: string;
  axes: RadarAxis[];
  is_reference: boolean;
}

/**
 * Multi-molecule radar comparison.
 */
export interface RadarComparison {
  profiles: RadarProfile[];
  reference: RadarProfile | null;
}

// =============================================================================
// Core Types
// =============================================================================

/**
 * Basic molecule information.
 */
export interface ScoringMoleculeInfo {
  input_string: string;
  canonical_smiles: string | null;
  molecular_formula: string | null;
  molecular_weight: number | null;
}

/**
 * Available scoring types.
 */
export type ScoringType =
  | 'ml_readiness'
  | 'np_likeness'
  | 'scaffold'
  | 'druglikeness'
  | 'safety_filters'
  | 'admet'
  | 'aggregator'
  | 'consensus'
  | 'lead_likeness'
  | 'salt_inventory'
  | 'ligand_efficiency'
  | 'tpsa_breakdown'
  | 'logp_breakdown'
  | 'bertz_detail'
  | 'fsp3_detail'
  | 'np_breakdown'
  | 'bioavailability_radar'
  | 'boiled_egg';

/**
 * Request for molecule scoring.
 */
export interface ScoringRequest {
  molecule: string;
  format?: 'auto' | 'smiles' | 'inchi' | 'mol';
  include?: ScoringType[];
}

/**
 * Response containing scoring results.
 */
export interface ScoringResponse {
  status: string;
  molecule_info: ScoringMoleculeInfo;
  ml_readiness: MLReadinessResult | null;
  np_likeness: NPLikenessResult | null;
  scaffold: ScaffoldResult | null;
  druglikeness: DrugLikenessResult | null;
  safety_filters: SafetyFilterResult | null;
  admet: ADMETResult | null;
  aggregator: AggregatorLikelihoodResult | null;
  consensus: ConsensusScore | null;
  lead_likeness: LeadLikeness | null;
  salt_inventory: SaltInventory | null;
  ligand_efficiency: LigandEfficiency | null;
  tpsa_breakdown: TPSABreakdown | null;
  logp_breakdown: LogPBreakdown | null;
  bertz_detail: BertzDetail | null;
  fsp3_detail: Fsp3Detail | null;
  np_breakdown: NPBreakdown | null;
  bioavailability_radar: BioavailabilityRadar | null;
  boiled_egg: BoiledEgg | null;
  execution_time_ms: number;
}

/**
 * Scoring error response.
 */
export interface ScoringError {
  error: string;
  details?: {
    errors?: string[];
    warnings?: string[];
    format_detected?: string;
  };
}
